File: sdk/typescript/src/bunqueue/retry.ts
/**
 * Bunqueue Simple Mode — advanced in-process retry with backoff strategies.
 * 1:1 port of src/client/bunqueue/retry.ts (identical formulas).
 */

import type { RetryConfig, RetryStrategy } from './types.js';

/** Calculate backoff delay based on strategy. */
export function calculateBackoff(
  strategy: RetryStrategy,
  attempt: number,
  baseDelay: number,
  error: Error,
  config: RetryConfig
): number {
  switch (strategy) {
    case 'fixed':
      return baseDelay;

    case 'exponential':
      return baseDelay * 2 ** (attempt - 1);

    case 'jitter': {
      const exp = baseDelay * 2 ** (attempt - 1);
      return Math.floor(exp * (0.5 + Math.random()));
    }

    case 'fibonacci': {
      let a = 1;
      let b = 1;
      for (let i = 0; i < attempt - 1; i++) {
        const next = a + b;
        a = b;
        b = next;
      }
      return baseDelay * b;
    }

    case 'custom':
      if (config.customBackoff) {
        return config.customBackoff(attempt, error);
      }
      return baseDelay;

    default:
      return baseDelay;
  }
}

/** Execute a function with retry logic (the job stays active throughout). */
export function executeWithRetry<R>(fn: () => Promise<R>, config: RetryConfig): Promise<R> {
  const maxAttempts = config.maxAttempts ?? 3;
  const baseDelay = config.delay ?? 1000;
  const strategy = config.strategy ?? 'exponential';

  const attempt = (n: number): Promise<R> => {
    return fn().catch((err: unknown) => {
      const error = err instanceof Error ? err : new Error(String(err));
      if (n >= maxAttempts) throw error;
      if (config.retryIf && !config.retryIf(error, n)) throw error;

      const delay = calculateBackoff(strategy, n, baseDelay, error, config);
      return new Promise<R>((resolve) => {
        setTimeout(() => {
          resolve(attempt(n + 1));
        }, delay);
      });
    });
  };

  return attempt(1);
}

File: sdk/python/bunqueue/simple/retry.py
"""Advanced in-process retry with backoff strategies (mirror of retry.ts).

This retries INSIDE the processor while the job stays active (the worker's
lock heartbeats keep it owned) — different from the server-side
attempts/backoff, which re-queues the job.
"""

from __future__ import annotations

import random
import time
from typing import Any, Callable, Dict, TypeVar

R = TypeVar("R")

STRATEGIES = ("fixed", "exponential", "jitter", "fibonacci", "custom")


def calculate_backoff(
    strategy: str,
    attempt: int,
    base_delay: float,
    error: BaseException,
    config: Dict[str, Any],
) -> float:
    """Delay in ms for the given attempt (1-based), same formulas as retry.ts."""
    if strategy == "fixed":
        return base_delay
    if strategy == "exponential":
        return base_delay * (2 ** (attempt - 1))
    if strategy == "jitter":
        exp = base_delay * (2 ** (attempt - 1))
        return float(int(exp * (0.5 + random.random())))
    if strategy == "fibonacci":
        a, b = 1, 1
        for _ in range(attempt - 1):
            a, b = b, a + b
        return base_delay * b
    if strategy == "custom":
        custom = config.get("custom_backoff") or config.get("customBackoff")
        if custom:
            return float(custom(attempt, error))
        return base_delay
    return base_delay


def execute_with_retry(fn: Callable[[], R], config: Dict[str, Any]) -> R:
    """Run ``fn`` retrying on exception, same semantics as executeWithRetry."""
    max_attempts = int(config.get("max_attempts") or config.get("maxAttempts") or 3)
    base_delay = float(config.get("delay") or 1000)
    strategy = str(config.get("strategy") or "exponential")
    retry_if = config.get("retry_if") or config.get("retryIf")

    attempt = 1
    while True:
        try:
            return fn()
        except BaseException as exc:  # noqa: BLE001 - retry semantics need broad catch
            if attempt >= max_attempts:
                raise
            if retry_if is not None and not retry_if(exc, attempt):
                raise
            delay_ms = calculate_backoff(strategy, attempt, base_delay, exc, config)
            time.sleep(delay_ms / 1000.0)
            attempt += 1
